R/dummy_multi_choice.R

```diff
@@ -0,0 +1,307 @@
+      prefix = prefix,
+      keep_original_cols = keep_original_cols,
+      skip = skip,
+      id = id
+    )
+  )
+}
+
+step_dummy_multi_choice_new <-
+  function(terms, role, trained, threshold, levels, input, other, naming,
+           prefix,  keep_original_cols, skip, id) {
+    step(
+      subclass = "dummy_multi_choice",
+      terms = terms,
+      role = role,
+      trained = trained,
+      threshold = threshold,
+      levels = levels,
+      input = input,
+      other = other,
+      naming = naming,
+      prefix = prefix,
+      keep_original_cols = keep_original_cols,
+      skip = skip,
+      id = id
+    )
+  }
+
+#' @export
+prep.step_dummy_multi_choice <- function(x, training, info = NULL, ...) {
+  col_names <- eval_select_recipes(x$terms, training, info)
+
+  multi_dummy_check_type(training[, col_names])
+
+  levels <- purrr::map(training[, col_names], as.character)
+  levels <- unlist(levels)
+  levels <- levels[!is.na(levels)]
+  levels <- keep_levels(levels, x$threshold, other = x$other)
+
+  step_dummy_multi_choice_new(
+    terms = x$terms,
+    role = x$role,
+    trained = TRUE,
+    threshold = x$threshold,
+    levels = levels,
+    input = col_names,
+    other = x$other,
+    naming = x$naming,
+    prefix = x$prefix,
+    keep_original_cols = get_keep_original_cols(x),
+    skip = x$skip,
+    id = x$id
+  )
+}
+
+multi_dummy_check_type <- function(dat) {
+  is_good <- function(x) {
+    is.factor(x) | is.character(x) | all(is.na(x))
+  }
+
+  all_good <- vapply(dat, is_good, logical(1))
+  label <- "factor, character, or NA"
+  if (!all(all_good))
+    rlang::abort(
+      paste0(
+        "All columns selected for the step",
+        " should be ",
+        label)
+    )
+  invisible(all_good)
+}
+
+#' @export
+bake.step_dummy_multi_choice <- function(object, new_data, ...) {
+
+  col_names <- object$input
+
+  indicators <- multi_dummy(new_data[, col_names], object$levels)
+
+  prefix <- object$prefix %||% col_names[1]
+
+  used_lvl <- gsub(paste0("^", prefix), "", colnames(indicators))
+  colnames(indicators) <- object$naming(prefix, used_lvl)
+
+  new_data <- bind_cols(new_data, as_tibble(indicators))
+  keep_original_cols <- get_keep_original_cols(object)
+
+  if (!keep_original_cols) {
+    new_data <- new_data[, !(colnames(new_data) %in% col_names), drop = FALSE]
+  }
+
+  as_tibble(new_data)
+}
+
+multi_dummy <- function(x, y) {
+  row_id <- rep(seq_len(nrow(x)), times = ncol(x))
+  values <- unlist(purrr::map(x, as.character), use.names = FALSE)
+
+  if (y$collapse) {
+    values[(!values %in% y$keep) & !is.na(values)] <- y$other
+  }
+
+  row_id <- row_id[!is.na(values)]
+  values <- values[!is.na(values)]
+
+  values <- factor(values)
+
+  res <- Matrix::sparseMatrix(
+    i = row_id,
+    j = as.numeric(values),
+    dims = c(nrow(x), length(levels(values)))
+  )
+
+  colnames(res) <- levels(values)
+
+  as.matrix(res) %>%
+    as_tibble() %>%
+    mutate_all(as.integer)
```

Comment:
Do we have any concerns about adding a new use of a superseded function to the package at this point? We could coerce to integer when still in matrix form like `as.matrix(res) %>% apply(2, as.integer) %>% as_tibble()`?

Reply:
done!